### PR TITLE
PXD-2916: implement uniqueKeys

### PR DESCRIPTION
**Pull request by fantix**
Auto migration is in uc-cdis/datamodelutils#6.

uc-cdis/gdcdatamodel#12 replaced indexes on these columns with unique ones:

* `program.name`
* `project.code`

also added unique indexes for:

* `(project_id, submitter_id)` of other nodes

### Improvements

* added unique indexes for uniqueKeys

### Deployment changes

* Sheepdog will now create (or replace) the missing indexes during migration
* New unique indexes have been added to all nodes uc-cdis/gdcdatamodel#12

### CODE REVIEW

**Comment by PlanXCyborg** [comment]
Jenkins Build 1 : time taken 6 min 21 sec
          Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/1/display/redirect

**Comment by fantix** [comment]
tests to fix 😞

**Comment by PlanXCyborg** [comment]
Jenkins Build 2 : time taken 30 min
          Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/2/display/redirect

**Comment by PlanXCyborg** [comment]
This PR contains code that is not formatted correctly according to [`black`](https://github.com/ambv/black). Run `black` on your code before merging.

<details>
<summary>Expand the full diff to see formatting changes</summary>

```diff
--- sheepdog/transactions/upload/transaction.py
+++ blackened
@@ -180,16 +180,16 @@
             entity.flush_to_session()
         try:
             self.session.flush()
         except IntegrityError as e:
             # don't handle non-unique constraint errors
-            if 'duplicate key value violates unique constraint' not in e.message:
+            if "duplicate key value violates unique constraint" not in e.message:
                 raise
             values = VALUES_REGEXP.findall(e.message)
             if not values:
                 raise
-            values = [v.strip() for v in values[0].split(',')]
+            values = [v.strip() for v in values[0].split(",")]
             keys = KEYS_REGEXP.findall(e.message)
             if len(keys) == len(values):
                 values = dict(zip(keys, values))
                 entities = []
                 label = None
@@ -203,17 +203,18 @@
                         entities.append(en)
                         label = en.node.label
                 else:
                     for entity in entities:
                         entity.record_error(
-                            '{} with {} already exists in the GDC'
-                            .format(entity.node.label, values),
+                            "{} with {} already exists in the GDC".format(
+                                entity.node.label, values
+                            ),
                             keys=keys,
                         )
                     if entities:
                         raise HandledIntegrityError()
-            self.record_error('{} already exists in the GDC'.format(values))
+            self.record_error("{} already exists in the GDC".format(values))
             raise HandledIntegrityError()
 
     @property
     def status_code(self):
         """
```
</details></br>

This formatting comment was generated automatically by a script in [uc-cdis/wool](https://github.com/uc-cdis/wool).

**Comment by PlanXCyborg** [comment]
Jenkins Build 3 : time taken 15 min
          Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/3/display/redirect

**Comment by PlanXCyborg** [comment]
Jenkins Build 4 : time taken 17 min
          Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/4/display/redirect

**Comment by PlanXCyborg** [comment]
Jenkins Build 5 : time taken 38 min
          Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/5/display/redirect

**Comment by PlanXCyborg** [comment]
Jenkins Build 6 : time taken 32 min
          Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/6/display/redirect

**Comment by PlanXCyborg** [comment]
Jenkins Build 7 : time taken 12 min
          Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/7/display/redirect

**Comment by PlanXCyborg** [comment]
Jenkins Build 8 : time taken 31 min
          Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/8/display/redirect

**Comment by PlanXCyborg** [comment]
Jenkins Build 9 : time taken 29 min
          Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/9/display/redirect

**Comment by PlanXCyborg** [comment]
Jenkins Build 10 : time taken 35 min
          Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/10/display/redirect

**Comment by PlanXCyborg** [comment]
Jenkins Build 11 : time taken 13 min
          Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/11/display/redirect

**Comment by fantix** [comment]
PyLint bug fixed in PyCQA/pylint#2760

**Comment by PlanXCyborg** [comment]
Jenkins Build 15 : time taken 27 min
          Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/15/display/redirect

**Comment by fantix** [comment]
Closing in favor of #266

**Comment by PlanXCyborg** [comment]
Jenkins Build 1 : time taken 45 min
Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/1/display/redirect


Test results:
	Passed: 109, Failed: 0, Skipped: 0



| Test | Time (`PR`) | Time (`master`) | Diff |
| - | - | - | - |
| @ExportPerformanceTests @Performance: Export all nodes of type: aliquot | 0.302 | 0.274 | 0.03 |
| @ExportPerformanceTests @Performance: Export all nodes of type: case | 0.354 | 0.222 | 0.13 |
| @ExportPerformanceTests @Performance: Export all nodes of type: experiment | 0.304 | 0.261 | 0.04 |
| @ExportPerformanceTests @Performance: Export all nodes of type: project | 0.083 | 0.042 | 0.04 |
| @ExportPerformanceTests @Performance: Export all nodes of type: read_group | 0.509 | 0.428 | 0.08 |
| @ExportPerformanceTests @Performance: Export all nodes of type: sample | 0.444 | 0.359 | 0.09 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: aliquot | 0.174 | 0.146 | 0.03 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: case | 0.238 | 0.214 | 0.02 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: experiment | 0.404 | 0.174 | 0.23 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: read_group_qc | 0.072 | 0.056 | 0.02 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: sample | 0.102 | 0.094 | 0.01 |
| @SubmissionPerformanceTests @Performance: Submission demographic 1000 | 30.078 | 26.739 | 3.34 |
| @SubmissionPerformanceTests @Performance: Submission publication 1000 | 20.839 | 19.284 | 1.56 |

**Comment by PlanXCyborg** [comment]
Jenkins Build 2 : time taken 37 min
Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/2/display/redirect


Test results:
	Passed: 109, Failed: 0, Skipped: 0



| Test | Time (`PR`) | Time (`master`) | Diff |
| - | - | - | - |
| @ExportPerformanceTests @Performance: Export all nodes of type: aliquot | 0.296 | 0.274 | 0.02 |
| @ExportPerformanceTests @Performance: Export all nodes of type: experiment | 0.397 | 0.261 | 0.14 |
| @ExportPerformanceTests @Performance: Export all nodes of type: program | 0.186 | 0.068 | 0.12 |
| @ExportPerformanceTests @Performance: Export all nodes of type: project | 0.046 | 0.042 | 0.00 |
| @ExportPerformanceTests @Performance: Export all nodes of type: read_group_qc | 0.194 | 0.084 | 0.11 |
| @ExportPerformanceTests @Performance: Export all nodes of type: sample | 0.492 | 0.359 | 0.13 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: aliquot | 0.282 | 0.146 | 0.14 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: case | 0.277 | 0.214 | 0.06 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: experiment | 0.201 | 0.174 | 0.03 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: read_group | 0.165 | 0.064 | 0.10 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: read_group_qc | 0.131 | 0.056 | 0.07 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: sample | 0.278 | 0.094 | 0.18 |
| @QueryPerformanceTests @Performance: Executing bottomUp query # 0 | 0.884 | 0.615 | 0.27 |
| @QueryPerformanceTests @Performance: Executing topDown query # 0 | 0.168 | 0.072 | 0.10 |

**Comment by PlanXCyborg** [comment]
Jenkins Build 3 : time taken 49 min
Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/3/display/redirect


Test results:
	Passed: 109, Failed: 0, Skipped: 0



| Test | Time (`PR`) | Time (`master`) | Diff |
| - | - | - | - |
| @ExportPerformanceTests @Performance: Export all nodes of type: aliquot | 0.296 | 0.274 | 0.02 |
| @ExportPerformanceTests @Performance: Export all nodes of type: experiment | 0.397 | 0.261 | 0.14 |
| @ExportPerformanceTests @Performance: Export all nodes of type: program | 0.186 | 0.068 | 0.12 |
| @ExportPerformanceTests @Performance: Export all nodes of type: project | 0.046 | 0.042 | 0.00 |
| @ExportPerformanceTests @Performance: Export all nodes of type: read_group_qc | 0.194 | 0.084 | 0.11 |
| @ExportPerformanceTests @Performance: Export all nodes of type: sample | 0.492 | 0.359 | 0.13 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: aliquot | 0.282 | 0.146 | 0.14 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: case | 0.277 | 0.214 | 0.06 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: experiment | 0.201 | 0.174 | 0.03 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: read_group | 0.165 | 0.064 | 0.10 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: read_group_qc | 0.131 | 0.056 | 0.07 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: sample | 0.278 | 0.094 | 0.18 |
| @QueryPerformanceTests @Performance: Executing bottomUp query # 0 | 0.884 | 0.615 | 0.27 |
| @QueryPerformanceTests @Performance: Executing topDown query # 0 | 0.168 | 0.072 | 0.10 |

**Comment by PlanXCyborg** [comment]
Jenkins Build 5 : time taken 1 hr 13 min
Check the https://jenkins.planx-pla.net/job/perf-sheepdog/job/PR-259/5/display/redirect


Test results:
	Passed: 109, Failed: 0, Skipped: 0



| Test | Time (`PR`) | Time (`master`) | Diff |
| - | - | - | - |
| @ExportPerformanceTests @Performance: Export all nodes of type: experiment | 0.276 | 0.261 | 0.01 |
| @ExportPerformanceTests @Performance: Export all nodes of type: project | 0.056 | 0.042 | 0.01 |
| @ExportPerformanceTests @Performance: Export all nodes of type: read_group | 0.635 | 0.428 | 0.21 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: case | 0.23 | 0.214 | 0.02 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: experiment | 0.226 | 0.174 | 0.05 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: read_group | 0.109 | 0.064 | 0.04 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: read_group_qc | 0.063 | 0.056 | 0.01 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: sample | 0.11 | 0.094 | 0.02 |
| @SubmissionPerformanceTests @Performance: Submission aligned_reads_index 100 | 17.817 | 16.87 | 0.95 |
| @SubmissionPerformanceTests @Performance: Submission case 100 | 2.395 | 2.235 | 0.16 |
| @SubmissionPerformanceTests @Performance: Submission demographic 100 | 2.605 | 2.254 | 0.35 |
| @SubmissionPerformanceTests @Performance: Submission experimental_metadata 100 | 18.217 | 16.176 | 2.04 |
| @SubmissionPerformanceTests @Performance: Submission exposure 10 | 0.388 | 0.367 | 0.02 |
| @SubmissionPerformanceTests @Performance: Submission exposure 100 | 2.279 | 2.169 | 0.11 |
| @SubmissionPerformanceTests @Performance: Submission family_history 100 | 2.638 | 2.018 | 0.62 |
| @SubmissionPerformanceTests @Performance: Submission keyword 100 | 2.167 | 1.94 | 0.23 |
| @SubmissionPerformanceTests @Performance: Submission publication 1000 | 22.626 | 19.284 | 3.34 |
| @SubmissionPerformanceTests @Performance: Submission read_group_qc 100 | 3.959 | 3.603 | 0.36 |
| @SubmissionPerformanceTests @Performance: Submission slide_image 10 | 1.913 | 1.796 | 0.12 |
| @SubmissionPerformanceTests @Performance: Submission slide_image 100 | 17.47 | 14.74 | 2.73 |
| @SubmissionPerformanceTests @Performance: Submission submitted_copy_number 100 | 17.301 | 16.181 | 1.12 |
| @SubmissionPerformanceTests @Performance: Submission submitted_methylation 100 | 20.026 | 17.149 | 2.88 |
| @SubmissionPerformanceTests @Performance: Submission submitted_somatic_mutation 100 | 17.699 | 15.312 | 2.39 |
| @SubmissionPerformanceTests @Performance: Submission submitted_unaligned_reads 100 | 17.862 | 16.488 | 1.37 |

File: sheepdog/transactions/upload/transaction.py [hunk 1]
```diff
@@ -172,7 +178,43 @@ def flush(self):
...
+                    else:
+                        if label and label != en.node.label:
+                            break
+                        entities.append(en)
+                        label = en.node.label
+                else:  # pylint: disable=useless-else-on-loop
+                    # https://github.com/PyCQA/pylint/pull/2760
+                    for entity in entities:
+                        entity.record_error(
+                            "{} with {} already exists in the GDC".format(
```
Comment (zflamig) [inline sheepdog/transactions/upload/transaction.py]: while we are here, can we change this to just be `{} with {} already exists` so its more agnostic?

**Pull request merged**